adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -882,15 +880,51 @@ void handlePublishedMessage(final MqttPublishMessage message, final MqttEndpoint
                             onMessageUndeliverable(context);
                         }
                         TracingHelper.logError(span, processing.cause());
-                        if (context.deviceEndpoint().isConnected()) {
-                            span.log("closing connection to device");
-                            context.deviceEndpoint().close();
-                        }
+                        handleMqttConnectionOnError(context, processing.cause(), span);
                     }
                     span.finish();
                 });
     }
 
+    private void handleMqttConnectionOnError(
+            final MqttContext context,
+            final Throwable cause,
+            final Span span) {
+        final String tenantId = context.tenant();
+        final MqttEndpoint endpoint = context.deviceEndpoint();
+
+        if (endpoint.isConnected()) {
+            if (Objects.isNull(tenantId)) {
+                span.log("closing connection to device");
+                endpoint.close();
+            } else {
+                getTenantConfiguration(tenantId, span.context())
+                        .onSuccess(tenantObject -> {
+                            // Close the MQTT connection, if the flag indicating that the connection should be closed
+                            // on occurrence of error is set to true.
+                            // Also close the connection, even though the above flag is set to false provided the errors
+                            // are due to the following reasons:
+                            // - The tenant is not registered.
+                            // - The tenant or adapter is disabled
+                            // - The device is disabled or not registered.
+                            if (tenantObject.isCloseMqttConnectionOnError()
+                                    || cause instanceof AdapterDisabledException
+                                    || cause instanceof RegistrationAssertionException
+                                    || cause instanceof TenantNotRegisteredException) {
+                                span.log("closing connection to device");
+                                endpoint.close();
```

Comment:
@kaniyan Regarding the `RegistrationAssertionException` used here:
I think in case an MQTT gateway has sent a message on behalf of a device, and that device doesn't exist anymore (hence a `RegistrationAssertionException` is thrown), the MQTT connection to the gateway shouldn't be closed